cr-examples/samples/src/main/java/oracle/code/samples/MathOptimizer.java

```diff
@@ -0,0 +1,281 @@
+        codeModel = codeModel.transform(OpTransformer.LOWERING_TRANSFORMER);
+        System.out.println("After Lowering: ");
+        System.out.println(codeModel.toText());
+
+        System.out.println("\nEvaluate");
+        // The Interpreter Invoke should launch new exceptions
+        var result = Interpreter.invoke(MethodHandles.lookup(), codeModel, 10);
+        System.out.println(result);
+
+        // Select invocation calls and display the lines
+        System.out.println("\nPlaying with Traverse");
+        codeModel.traverse(null, (map, op) -> {
+            if (op instanceof JavaOp.InvokeOp invokeOp) {
+                System.out.println("Function Name: " + invokeOp.invokeDescriptor().name());
+
+                // Maybe Location should throw a new exception instead of the NPE,
+                // since it is possible we don't have a location after a transformation has been done.
+                Location location = invokeOp.location();
+                if (location != null) {
+                    int line = location.line();
+                    System.out.println("Line " + line);
+                    System.out.println("Class: " + invokeOp.getClass());
+                    // Detect Math::pow
+                    boolean contains = invokeOp.invokeDescriptor().equals(JAVA_LANG_MATH_POW);
+                    if (contains) {
+                        System.out.println("Method: " + invokeOp.invokeDescriptor().name());
+                    }
+                } else {
+                    System.out.println("[WARNING] Location is null");
+                }
+            }
+            return map;
+        });
+    }
+
+    // Goal: obtain and check the value of the function parameters.
+    // The function inspectParameterRecursive implements this method
+    // in a much simpler and shorter manner. We can keep this first
+    // implementation as a reference.
+    private static boolean inspectParameter(Value operand, final int value) {
+        final Boolean[] isMultipliedByTwo = new Boolean[] { false };
+        if (operand instanceof Op.Result res) {
+            if (res.op() instanceof JavaOp.ConvOp convOp) {
+                convOp.operands().forEach(v -> {
+                    if (v instanceof Op.Result res2) {
+                        if (res2.op() instanceof CoreOp.ConstantOp constantOp) {
+                            if (constantOp.value() instanceof Integer parameter) {
+                                if (parameter.intValue() == value) {
+                                    // Transformation is valid
+                                    isMultipliedByTwo[0] = true;
+                                }
+                            }
+                        }
+                    }
+                });
+            }
+        }
+        return isMultipliedByTwo[0];
+    }
+
+    // Inspect a value for a parameter
+    static boolean inspectParameterRecursive(JavaOp.ConvOp convOp, int valToMatch) {
```

Comment:
At some point we will get better at pattern matching, we need the language enhancements for pattern matching.

In the interim see the experiment in `jdk.incubator.code.analysis.Patterns` and its use [here](https://github.com/openjdk/babylon/blob/code-reflection/test/jdk/java/lang/reflect/code/ad/ExpressionElimination.java#L58). Not suggesting you use it here, just for information which you might find interesting.